File: scripts_go/CPCTF_Very_Exciting.go
package main

import (
	"bufio"
	"bytes"
	"encoding/hex"
	"flag"
	"fmt"
	"net"
	"regexp"
	"time"
)

// 题目利用点：
// 1. 服务先打印 flag 在 (secret_key, exciting_iv) 下的流密码密文；
// 2. 随后又开放“自选 IV + 自选明文”的加密 oracle，且仍然复用同一个 secret_key；
// 3. 如果把 IV 设成同一个 exciting_iv，并发送等长全 0 明文，返回值就会等于 keystream；
// 4. 再把 keystream 与 flag 密文异或，即可恢复原始 flag。

var (
	ivPattern         = regexp.MustCompile(`exciting_iv[^:：]*[:：]\s*([0-9a-f]+)`)
	ciphertextPattern = regexp.MustCompile(`=>\s*([0-9a-f]+)`)
)

func readUntilAny(reader *bufio.Reader, conn net.Conn, markers [][]byte) ([]byte, error) {
	var buffer bytes.Buffer
	for {
		data := buffer.Bytes()
		for _, marker := range markers {
			if bytes.Contains(data, marker) {
				return append([]byte(nil), data...), nil
			}
		}

		if err := conn.SetReadDeadline(time.Now().Add(10 * time.Second)); err != nil {
			return nil, err
		}

		b, err := reader.ReadByte()
		if err != nil {
			return append([]byte(nil), buffer.Bytes()...), err
		}
		buffer.WriteByte(b)
	}
}

func extractIVAndCiphertext(text string) (string, []byte, error) {
	ivMatch := ivPattern.FindStringSubmatch(text)
	cipherMatch := ciphertextPattern.FindStringSubmatch(text)
	if len(ivMatch) != 2 || len(cipherMatch) != 2 {
		return "", nil, fmt.Errorf("无法解析 exciting_iv / exciting_flag")
	}

	ciphertext, err := hex.DecodeString(cipherMatch[1])
	if err != nil {
		return "", nil, fmt.Errorf("无法解析 flag 密文字节: %w", err)
	}
	return ivMatch[1], ciphertext, nil
}

func xorBytes(left, right []byte) ([]byte, error) {
	if len(left) != len(right) {
		return nil, fmt.Errorf("异或双方长度不一致")
	}
	out := make([]byte, len(left))
	for i := range left {
		out[i] = left[i] ^ right[i]
	}
	return out, nil
}

func zeroHex(byteLen int) string {
	result := make([]byte, byteLen*2)
	for i := range result {
		result[i] = '0'
	}
	return string(result)
}

func hexToBytes(hexText string) ([]byte, error) {
	result, err := hex.DecodeString(hexText)
	if err != nil {
		return nil, fmt.Errorf("无法解析十六进制: %w", err)
	}
	return result, nil
}

func recoverFlag(host string, port int) (string, error) {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", host, port), 10*time.Second)
	if err != nil {
		return "", err
	}
	defer conn.Close()

	reader := bufio.NewReader(conn)
	initialMarkers := [][]byte{
		[]byte("Enter your boring 'favorite' (Hex): "),
		[]byte("输入你无聊的“最爱”（十六进制）: "),
	}
	secondMarkers := [][]byte{
		[]byte("Enter your own 'very_exciting' IV (Hex): "),
		[]byte("Enter your own 'very_exciting' IV (Hex):"),
	}

	banner, err := readUntilAny(reader, conn, initialMarkers)
	if err != nil {
		return "", err
	}

	excitingIVHex, excitingFlag, err := extractIVAndCiphertext(string(banner))
	if err != nil {
		return "", err
	}

	if _, err = fmt.Fprintf(conn, "%s\n", zeroHex(len(excitingFlag))); err != nil {
		return "", err
	}

	if _, err = readUntilAny(reader, conn, secondMarkers); err != nil {
		return "", err
	}
	if _, err = fmt.Fprintf(conn, "%s\n", excitingIVHex); err != nil {
		return "", err
	}

	var response bytes.Buffer
	for {
		if err = conn.SetReadDeadline(time.Now().Add(5 * time.Second)); err != nil {
			return "", err
		}
		chunk := make([]byte, 4096)
		n, readErr := reader.Read(chunk)
		if n > 0 {
			response.Write(chunk[:n])
		}
		if readErr != nil {
			netErr, ok := readErr.(net.Error)
			if ok && netErr.Timeout() {
				break
			}
			break
		}
	}

	keystreamMatch := ciphertextPattern.FindStringSubmatch(response.String())
	if len(keystreamMatch) != 2 {
		return "", fmt.Errorf("无法解析 keystream")
	}
	keystream, err := hexToBytes(keystreamMatch[1])
	if err != nil {
		return "", err
	}

	flagBytes, err := xorBytes(excitingFlag, keystream)
	if err != nil {
		return "", err
	}
	return string(flagBytes), nil
}

func main() {
	host := flag.String("host", "133.88.122.244", "challenge host")
	port := flag.Int("port", 32007, "challenge port")
	flag.Parse()

	result, err := recoverFlag(*host, *port)
	if err != nil {
		panic(err)
	}
	fmt.Println(result)
}

File: scripts_python/CPCTF_Very_Exciting.py
"""
CPCTF - Very Exciting

思路：服务先打印 flag 在某个 (secret_key, exciting_iv) 下的流密码密文，
随后又提供一个“自选 IV + 自选明文”的加密 oracle，而且仍然沿用同一个 secret_key。
只要把用户提交的 IV 设成打印出来的 exciting_iv，并提交与 flag 密文等长的全 0 明文，
服务返回的密文就会退化成裸露的 keystream；再与 flag 密文异或即可恢复 flag。
"""

from __future__ import annotations

import argparse
import re
import socket
from typing import Iterable

# 兼容附件中的中文提示与远端服务中的英文提示。
INITIAL_PROMPTS = [
    b"Enter your boring 'favorite' (Hex): ",
    "输入你无聊的“最爱”（十六进制）: ".encode(),
]
SECOND_PROMPTS = [
    b"Enter your own 'very_exciting' IV (Hex): ",
    b"Enter your own 'very_exciting' IV (Hex):",
]
IV_PATTERN = re.compile(r"exciting_iv[^:：]*[:：]\s*([0-9a-f]+)")
CIPHERTEXT_PATTERN = re.compile(r"=>\s*([0-9a-f]+)")


def recv_until_any(sock: socket.socket, markers: Iterable[bytes]) -> bytes:
    """持续接收数据，直到缓冲区中出现任意一个目标提示。"""
    data = b""
    markers = tuple(markers)
    while True:
        if any(marker in data for marker in markers):
            return data
        chunk = sock.recv(4096)
        if not chunk:
            return data
        data += chunk


def extract_iv_and_ciphertext(banner: str) -> tuple[str, bytes]:
    """从服务横幅中提取 IV 和 flag 密文。"""
    iv_match = IV_PATTERN.search(banner)
    ciphertext_match = CIPHERTEXT_PATTERN.search(banner)
    if iv_match is None or ciphertext_match is None:
        raise ValueError("无法从服务输出中解析 exciting_iv / exciting_flag")
    return iv_match.group(1), bytes.fromhex(ciphertext_match.group(1))


def xor_bytes(left: bytes, right: bytes) -> bytes:
    """逐字节异或两个等长字节串。"""
    if len(left) != len(right):
        raise ValueError("异或双方长度必须一致")
    return bytes(a ^ b for a, b in zip(left, right))


def recover_flag(host: str, port: int) -> str:
    """连接远端服务并自动恢复 flag。"""
    with socket.create_connection((host, port), timeout=10) as sock:
        sock.settimeout(10)

        # 第一步：读到首个输入提示，解析 flag 密文与对应 IV。
        banner = recv_until_any(sock, INITIAL_PROMPTS).decode(errors="replace")
        exciting_iv_hex, exciting_flag = extract_iv_and_ciphertext(banner)

        # 第二步：发送与 flag 密文等长的全 0 明文。
        zero_plaintext_hex = "00" * len(exciting_flag)
        sock.sendall((zero_plaintext_hex + "\n").encode())

        # 第三步：读到 IV 提示，并复用服务先前打印出来的同一个 IV。
        recv_until_any(sock, SECOND_PROMPTS)
        sock.sendall((exciting_iv_hex + "\n").encode())

        # 第四步：读取后续输出。返回的“密文”此时其实就是 keystream。
        response = b""
        while True:
            try:
                chunk = sock.recv(4096)
            except socket.timeout:
                break
            if not chunk:
                break
            response += chunk

    response_text = response.decode(errors="replace")
    keystream_match = CIPHERTEXT_PATTERN.search(response_text)
    if keystream_match is None:
        raise ValueError("无法从二次响应中解析 keystream")

    keystream = bytes.fromhex(keystream_match.group(1))
    flag_bytes = xor_bytes(exciting_flag, keystream)
    return flag_bytes.decode()


def main() -> None:
    parser = argparse.ArgumentParser(description="Solve CPCTF Very Exciting via IV/keystream reuse")
    parser.add_argument("--host", default="133.88.122.244", help="challenge host")
    parser.add_argument("--port", type=int, default=32007, help="challenge port")
    args = parser.parse_args()

    print(recover_flag(args.host, args.port))


if __name__ == "__main__":
    main()
